dill/_dill.py

```diff
@@ -410,9 +416,69 @@ def save_numpy_array(pickler, obj):
             raise PicklingError(msg)
         logger.trace_setup(self)
         StockPickler.dump(self, obj)
-
     dump.__doc__ = StockPickler.dump.__doc__
 
+    def save(self, obj, save_persistent_id=True, *, name=None):
+        """If self._refonfail is True, try to save object by reference if pickling fails."""
+        if not self._refonfail:
+            super().save(obj, save_persistent_id)
+            return
+        # Disable framing (right after the framer.init_framing() call at dump()).
+        self.framer.current_frame = None
+        # Store initial state.
+        position = self._file_tell()
+        memo_size = len(self.memo)
+        try:
+            super().save(obj, save_persistent_id)
+        except UNPICKLEABLE_ERRORS + (AttributeError,) as error_stack:
+            # AttributeError may happen in the save_global() call from a child object.
+            if (type(error_stack) == AttributeError
+                    and "no attribute '__name__'" not in error_stack.args[0]):
+                raise
+            # Roll back the stream.
+            self._file_seek(position)
+            self._file_truncate()
+            # Roll back memo.
+            for _ in range(len(self.memo) - memo_size):
+                self.memo.popitem()  # LIFO order is guaranteed since 3.7
+            try:
+                if isinstance(obj, ModuleType) and \
+                        (_is_builtin_module(obj) or obj is sys.modules['dill']):
+                    self.save_reduce(_import_module, (obj.__name__,), obj=obj)
+                else:
+                    self.save_global(obj, name)
+            except (AttributeError, PicklingError) as error:
+                if getattr(self, '_trace_stack', None) and id(obj) == self._trace_stack[-1]:
+                    # Roll back trace state.
+                    self._trace_stack.pop()
+                    self._size_stack.pop()
+                raise error from error_stack
+            logger.trace(self, "# X: fallback to save_global: <%s object at %#012x>",
+                         type(obj).__name__, id(obj), obj=obj)
+
+    def _save_module_dict(self, obj):
+        """
+        Use object name in the module namespace as a last resource to try to
```

Comment:
Did you mean: "last resort" or "last resource"?  Again, you might want to be more specific about "when pickling fails".

Reply:
I meant "last resort". There's not much specificity here... What I could add is

- When pickling an object raises an exception and the `save_global()` fallback call can't find the object by itself.

Reply:
Just rewrote everything...